wire-ios-request-strategy/Sources/Message Sending/ProteusMessagePayloadBuilder.swift

```diff
@@ -0,0 +1,166 @@
+        }
+    }
+
+    private func unQualifiedData(messageInfo: MessageInfo, encryptedDatas: [String: Data], externalData: Data? = nil) async throws -> Data {
+        var userEntries = [Proteus_UserEntry]()
+        for (_, entries) in messageInfo.listClients {
+
+            for (userId, userClientDatas) in entries {
+
+                let userEntry = proteusUserEntry(userClientDatas: userClientDatas, for: userId, encryptedDatas: encryptedDatas)
+                userEntries.append(userEntry)
+            }
+        }
+
+        let message = Proteus_NewOtrMessage(
+            withSenderId: messageInfo.selfClientID.hexRemoteIdentifier,
+            nativePush: messageInfo.nativePush,
+            recipients: userEntries,
+            missingClientsStrategy: messageInfo.missingClientsStrategy,
+            blob: externalData
+        )
+
+        return try message.serializedData()
+    }
+
+    private func qualifiedData(messageInfo: MessageInfo, encryptedDatas: [String: Data], externalData: Data? = nil) async throws -> Data {
+
+        var finalRecipients = [Proteus_QualifiedUserEntry]()
+        for (domain, entries) in messageInfo.listClients {
+
+            var userEntries = [Proteus_UserEntry]()
+            for (userId, userClientDatas) in entries {
+
+                let userEntry = proteusUserEntry(userClientDatas: userClientDatas, for: userId, encryptedDatas: encryptedDatas)
+                userEntries.append(userEntry)
+            }
+
+            finalRecipients.append(
+                Proteus_QualifiedUserEntry(withDomain: domain, userEntries: userEntries)
+            )
+        }
+
+        let message = Proteus_QualifiedNewOtrMessage(
+            withSenderId: messageInfo.selfClientID.hexRemoteIdentifier,
+            nativePush: messageInfo.nativePush,
+            recipients: finalRecipients,
+            missingClientsStrategy: messageInfo.missingClientsStrategy,
+            blob: externalData
+        )
+
+        return try message.serializedData()
+    }
+
+    private func proteusUserEntry(userClientDatas: [UserClientData],
+                                  for userID: UUID,
+                                  encryptedDatas: [String: Data]) -> Proteus_UserEntry {
+        let proteusUserID = Proteus_UserId.with({ $0.uuid = userID.uuidData })
+
+        let clientEntries = userClientDatas.compactMap { userClientData in
+            let clientId = Proteus_ClientId.with({ $0.client = userClientData.sessionID.clientID.hexRemoteIdentifier })
+
+            if let data = userClientData.data {
+                return Proteus_ClientEntry(withClientId: clientId, data: data)
+            }
+            if let encryptedData = encryptedDatas[userClientData.sessionID.rawValue] {
+                return Proteus_ClientEntry(withClientId: clientId, data: encryptedData)
+            } else {
+                assertionFailure("should not happen")
```

Comment:
for some reasons, this does happen, giving X proteusSessions to encryptBatch and receiving only Y [String: Data] values. Y< X.

Reply:
actually it's when session is not establish, and it's normal as missing clients will be handled in MessageSender